kvstoremesh/metrics/metrics.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/collectors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metrics")
+)
+
+var Cell = cell.Module(
+	"kvstoremesh-metrics",
+	"KVStoreMesh metrics",
+
+	cell.Config(MetricsConfig{}),
+	cell.Invoke(registerMetricsManager),
+)
+
+type MetricsConfig struct {
+	// PrometheusServeAddr IP:Port on which to serve prometheus metrics (pass ":Port" to bind on all interfaces, "" is off)
+	PrometheusServeAddr string
+}
+
+func (def MetricsConfig) Flags(flags *pflag.FlagSet) {
+	flags.String(option.PrometheusServeAddr, def.PrometheusServeAddr, "Address to serve Prometheus metrics")
+}
+
+type metricsManager struct {
+	registry *prometheus.Registry
+	server   http.Server
+
+	metrics []metric.WithMetadata
+}
+
+type params struct {
+	cell.In
+
+	MetricsConfig
+
+	Metrics []metric.WithMetadata `group:"hive-metrics"`
+}
+
+func registerMetricsManager(lc hive.Lifecycle, params params) error {
+	manager := metricsManager{
+		registry: prometheus.NewPedanticRegistry(),
+		server:   http.Server{Addr: params.PrometheusServeAddr},
+		metrics:  params.Metrics,
+	}
+
+	if params.PrometheusServeAddr != "" {
+		lc.Append(&manager)
+	} else {
+		log.Info("Prometheus metrics are disabled")
+	}
+
+	return nil
+}
+
+func (mm *metricsManager) Start(hive.HookContext) error {
+	log.Info("Registering metrics")
+
+	mm.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
```

Comment:
Should these be namespaced (i.e. `cilium_kvstoremesh`)?

Reply:
It seems that there is currently a bit of inconsistency in the code base, since process-related metrics are currently namespaced in the agents and cilium-operator, and non-namespaced in hubble relay and in the clustermesh-apiserver.  Go related metrics instead are always non-namespaced. Do you feel it would be better change the process ones to be namespaced in clustermesh-apiserver/kvstoremesh?

Reply:
>  So from what I can tell, with reuising the existing pkg/metrics declarations this will result in using the "cilium_" namespace?

As for the cilium specific metrics, the [namespace is changed to `cilium_kvstoremesh`](https://github.com/cilium/cilium/blob/96abcb95bc34356c9465de536d35749039419815/kvstoremesh/main.go#L45) (and `cilium_clustermesh_apiserver` in the clustermesh-apiserver case). The only metrics that are currently not namespaced in those two cases are the generic ones (go and process).

Reply:
ah I see, missed that line, thought this would be the same as the agent. Ok looks good thanks.